secrets/provider_registry.go

```diff
@@ -0,0 +1,131 @@
+package secrets
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/buildkite/agent/v3/bootstrap/shell"
+	"github.com/puzpuzpuz/xsync"
+)
+
+type ProviderRegistry struct {
+	config ProviderRegistryConfig
+	// We have the two maps here because we only want to initialise a provider if a secret using that provider is used
+	// We shouldn't boot up any secrets providers that won't get used
+	candidates *xsync.MapOf[string, providerCandidate] // Candidates technically doesn't need to be a sync map as it's never altered after initialization, but i've made it one just for symmetry
+	providers  *xsync.MapOf[string, Provider]
+}
+
+type ProviderRegistryConfig struct {
+	Shell *shell.Shell
```

Comment:
i don't love passing the bootstrap's shell into the provider registry, but it would be nice for it to be able to log things, and when it creates [`secrets.EnvSecret`s](https://github.com/buildkite/agent/pull/1848/files#diff-f427d9a2346cf79f74bece1e3483ad9d6c0cc7dad4cf7818de4db9f9658855f9R26), it needs to be able to pass the shell's env into them so that they can store themselves. This seems a bit nasty and i'd appreciate some C&C on how to make it a bit better